### PR TITLE
Send messages from scheduler to tasks

### PR DESCRIPTION
**Pull request by riga**
## Description

This PR enables the scheduler to send messages to particular tasks. More precisely, messages are sent to workers via the RPC message system which dispatches them to a specified task. Running tasks 
can access a `multiprocessing.Queue` object (named `scheduler_messages`, set via the `StatusReporter`) which stores incoming messages. Users can implement custom behavior to react to certain messages (see code example below).

The new config `scheduler.send_messages` controls whether a scheduler is capable of sending messages in the first place. By defining `accepted_messages` a task can define whether it accepts messages, independent of the scheduler the worker is talking to. The message button is only visible in the UI when the scheduler is able to send messages, and the task can receive them (see screenshots).

There is also a simple "respond" mechanism so tasks can immediately respond to incoming messages which is shown in the scheduler.

"Send message" action button:

Action button

Message prompt:

Message prompt

Awaiting response:

Awaiting response

Response arrived:

Response arrived

When `Await response` is not checked, the prompt closes upon `Send`. Otherwise, the prompt remains open and a response container is shown. This triggers a simple polling to fetch and display the actual response.

## Motivation and Context

There are many use cases when the messaging could be useful. One particular example is machine learning. Our training tasks publish their current results to the scheduler. Although they contain predefined termination criteria, it could be nice to manually trigger *graceful* training termination if the training stagnates:

```python
class Training(luigi.Task):

    def requires(self):
        ...

    def output(self):
        return luigi.LocalTarget("my_ml_model.pb")

    def run(self):
        # setup the training
        model = ...

        for _ in training_loop():
            # training stuff with internal termination criterion
            model.train(...)

            # check messages
            if not self.scheduler_messages.empty():
                msg = self.scheduler_messages.get()
                if msg.content == "terminate":
                    break
                else:
                    msg.respond("unknown message")

        # save the model
        model.save(self.output().path)
```

## Have you tested this? If so, how?

Yep, I added unit tests and docs.

### CODE REVIEW

File: luigi/worker.py [hunk 1]
```diff
@@ -1148,3 +1155,14 @@ def set_worker_processes(self, n):
 
         # tell the scheduler
         self._scheduler.add_worker(self._id, {'workers': self.worker_processes})
+
+    @rpc_message_callback
+    def dispatch_scheduler_message(self, task_id, message):
```
[the remark this answers is not in the document]
Comment (riga) [reply luigi/worker.py]: Implemented a draft if this reply/response mechanism.

File: luigi/scheduler.py [hunk 2]
```diff
@@ -932,6 +936,31 @@ def disable_worker(self, worker):
...
+    @rpc_method()
+    def send_scheduler_message(self, worker, task, message):
+        if not self._config.send_messages:
+            return {"messageId": None}
+
+        message_id = str(uuid.uuid4())
+        self._state.get_worker(worker).add_rpc_message('dispatch_scheduler_message', task_id=task,
+                                                       message_id=message_id, message=message)
+
+        return {"messageId": message_id}
```
Comment (Tarrasch) [inline luigi/scheduler.py]: yuck, I just realize that we mix camelCase and snake_case in the key-names. But it seems snake_case is more used. Maybe use that?

File: luigi/worker.py [hunk 3]
```diff
@@ -392,6 +420,10 @@ class worker(Config):
                                           description='If true, use multiprocessing also when '
                                           'running with 1 worker')
 
+    receive_messages = BoolParameter(default=True,
```
Comment (Tarrasch) [inline luigi/worker.py]: Maybe `can_receive_messages`?

Also I'm thinking of having the default to `False`, so the UI button doesn't appear by default.

File: luigi/worker.py [hunk 4]
```diff
@@ -447,17 +479,17 @@ def __init__(self, scheduler=None, worker_id=None, worker_processes=1, assistant
         if scheduler is None:
             scheduler = Scheduler()
 
+        self._config = worker(**kwargs)
+
+        assert self._config.wait_interval >= _WAIT_INTERVAL_EPS, "[worker] wait_interval must be positive"
+        assert self._config.wait_jitter >= 0.0, "[worker] wait_jitter must be equal or greater than zero"
+
```
Comment (Tarrasch) [inline luigi/worker.py]: Why the need to move this code?

Comment (Tarrasch) [reply luigi/worker.py]: (feel free to if you have a reason, I'm mostly curious)

File: luigi/worker.py [hunk 5]
```diff
@@ -540,7 +572,8 @@ def _generate_worker_info(self):
         # Generate as much info as possible about the worker
         # Some of these calls might not be available on all OS's
         args = [('salt', '%09d' % random.randrange(0, 999999999)),
-                ('workers', self.worker_processes)]
+                ('workers', self.worker_processes),
+                ('receive_messages', self._config.receive_messages)]
```
Comment (Tarrasch) [inline luigi/worker.py]: Hmmm... wait, why is this a property of the worker...

Thinking about it I'm sure it makes sense. But I really would like to see some user docs.

File: luigi/scheduler.py [hunk 6]
```diff
@@ -932,6 +936,31 @@ def disable_worker(self, worker):
     def set_worker_processes(self, worker, n):
         self._state.get_worker(worker).add_rpc_message('set_worker_processes', n=n)
 
+    @rpc_method()
+    def send_scheduler_message(self, worker, task, message):
```
[the remark this answers is not in the document]
Comment (riga) [reply luigi/scheduler.py]: ✔︎

File: luigi/scheduler.py (hunk 2 above)
Comment (riga) [reply luigi/scheduler.py]: ✔︎

File: luigi/worker.py (hunk 4 above)
Comment (riga) [reply luigi/worker.py]: This was required as `_generate_worker_info()` needs to access `_config`.

File: luigi/worker.py (hunk 5 above)
Comment (riga) [reply luigi/worker.py]: Mh, but you have a point. Another alternative would be to have sth like a "can_receive_messages()" hook (defaulting to false) defined on a task. Actually this makes more sense as the worker should just dispatch and let tasks decide what to do with messages. Are you fine with that change?

Comment (Tarrasch) [reply luigi/worker.py]: Well. I think it would make sense to both have message for tasks and workers.

Worker messages could be "shut down asap", "shut down task x", "decrease num workers" (we already have that, don't we)? Tasks messages I imagine are mostly task-specific.

----

I'm not sure how to implement it. Anyway, I'm fine with any changes, I'm sure if you think it becomes better so will I. :)

File: doc/luigi_patterns.rst [hunk 7]
```diff
@@ -317,3 +317,39 @@ built-in solutions. In the case of you're dealing with a file system
...
+
+.. code-block:: python
+
+    class Example(luigi.Task):
+
+        # common task setup
+        ...
+
+        # configure the task to accept all incoming messages
+        accepted_messages = True
```
[the remark this answers is not in the document]
Comment (riga) [reply doc/luigi_patterns.rst]: ✔︎

**Comment by Tarrasch** [comment]
Are you happy with this now? Is it ready to merge? Is it reasonably tested?

**Comment by riga** [comment]
Yep, I'm happy with the PR. Docs and tests should be fine.

**Comment by Tarrasch** [comment]
Thanks!! This is so cool!
